### PR TITLE
Added support for tunneling and updated tunnel schema

### PR DESCRIPTION
**Pull request by scsmncao**
- added tunneldecaporch
 - changed orchdaemon and main to include tunneldecaporch
 - updated orchagent Makefile

### CODE REVIEW

File: doc/swss-schema.md [hunk 1]
```diff
@@ -279,6 +279,8 @@ and reflects the LAG ports into the redis under: `LAG_TABLE:<team0>:port`
     ecn_mode                = "copy_from_outer" / "standard" ;standard: Behavior defined in RFC 6040 section 4.2
```
Comment (lguohan) [inline doc/swss-schema.md]: see above example, add
;field						value

Comment (scsmncao) [inline doc/swss-schema.md]: updated!

File: orchagent/tunneldecaporch.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+    int length_of_tunnel_attrs = 4;
+
+    // adding tunnel attributes to array and writing to ASIC_DB
+    sai_attribute_t tunnel_attrs[length_of_tunnel_attrs];
+
+    // tunnel type (only ipinip for now)
+    tunnel_attrs[0].id = SAI_TUNNEL_ATTR_TYPE;
+    tunnel_attrs[0].value.s32 = SAI_TUNNEL_IPINIP;
+
```
Comment (lguohan) [inline orchagent/tunneldecaporch.cpp]: looks like you need to set SAI_TUNNEL_ATTR_OVERLAY_INTERFACE for tunnel decap use. Before doing that, you need to craete a LOOPBACK type router interface, like below.

intf_attr.id = SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID;
intf_attr.value.oid = default_virtual_router_id;
intf_attrs.push_back(intf_attr);
intf_attr.id = SAI_ROUTER_INTERFACE_ATTR_TYPE;
intf_attr.value.s32 = SAI_ROUTER_INTERFACE_TYPE_LOOPBACK;
intf_attrs.push_back(intf_attr);
create_router_interface(&underlay_if, intf_attrs.size(), intf_attrs.data());

File: orchagent/tunneldecaporch.cpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    SWSS_LOG_ENTER();
+}
+
+/**
+ * Function Description:
+ *    @brief reads from APP_DB and creates tunnel 
+ */
+void TunnelDecapOrch::doTask(Consumer& consumer)
+{
+    if (consumer.m_toSync.empty())
```
Comment (stcheng) [inline orchagent/tunneldecaporch.cpp]: these two lines are no longer needed.

File: orchagent/tunneldecaporch.cpp [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    // gathering the switch virtual router id (this isn't actually used in brcm_sai but required for create)
+    sai_attribute_t attr;
+    attr.id = SAI_SWITCH_ATTR_DEFAULT_VIRTUAL_ROUTER_ID;
+    status = sai_switch_api->get_switch_attribute(1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Fail to get switch virtual router ID %d", status);
+        exit(EXIT_FAILURE);
+    }
+    sai_object_id_t virtualRouterId = attr.value.oid;
```
Comment (stcheng) [inline orchagent/tunneldecaporch.cpp]: why not using the gVirtualRouterId that is already get in main.cpp?

**Comment by stcheng** [comment]
Looks pretty good to me.

File: orchagent/tunneldecaporch.cpp (hunk 4 above)
Comment (scsmncao) [inline orchagent/tunneldecaporch.cpp]: you're right! fixed!

File: orchagent/tunneldecaporch.cpp (hunk 3 above)
Comment (scsmncao) [inline orchagent/tunneldecaporch.cpp]: fixed!

File: orchagent/tunneldecaporch.cpp (hunk 2 above)
Comment (scsmncao) [inline orchagent/tunneldecaporch.cpp]: added! it's initialized in main.cpp

File: orchagent/main.cpp [hunk 5]
```diff
@@ -31,9 +31,12 @@ sai_next_hop_api_t*         sai_next_hop_api;
 sai_next_hop_group_api_t*   sai_next_hop_group_api;
 sai_route_api_t*            sai_route_api;
 sai_lag_api_t*              sai_lag_api;
+sai_tunnel_api_t*           sai_tunnel_api;
 
 map<string, string> gProfileMap;
 sai_object_id_t gVirtualRouterId;
+sai_object_id_t overlayIfId;
+sai_object_id_t underlayIfId;
```
Comment (stcheng) [inline orchagent/main.cpp]: So this is not used currently, right?

Comment (scsmncao) [inline orchagent/main.cpp]: Nope, not yet, should I remove it for now?

Comment (stcheng) [inline orchagent/main.cpp]: No problem. If it will be used later, you can just leave it there.

File: orchagent/main.cpp [hunk 6]
```diff
@@ -171,6 +176,25 @@ int main(int argc, char **argv)
...
 
+    // create the overlay router interface to create a LOOPBACK type router interface (decap)
+    sai_attribute_t intf_attrs[2];
+    intf_attrs[0].id = SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID;
+    intf_attrs[0].value.oid = gVirtualRouterId;
+    intf_attrs[1].id = SAI_ROUTER_INTERFACE_ATTR_TYPE;
+    intf_attrs[1].value.s32 = SAI_ROUTER_INTERFACE_TYPE_LOOPBACK;
+    sai_object_id_t overlay_if;
+
+    status = sai_router_intfs_api->create_router_interface(&overlay_if, 2, intf_attrs);
```
Comment (lguohan) [inline orchagent/main.cpp]: why not use overlayIf directly here?

File: swssconfig/sample/sample.json [hunk 7]
```diff
@@ -51,4 +51,14 @@
 			},
 			"OP": "SET"
 		}
-	]
\ No newline at end of file
+		{
```
Comment (lguohan) [inline swssconfig/sample/sample.json]: can you put this as a separate sample file.

for example, netbouncer.json

also for the example output, put it as a separate file netbouncer.json.output

File: swssconfig/sample/sample.json [hunk 8]
```diff
@@ -51,4 +51,14 @@
...
+			"QOS_TABLE:TUNNEL_DECAP_TABLE:NETBOUNCER" : {
+				"tunnel_type":"IPINIP",
+				"dst_ip":"127.0.0.1",
```
Comment (lguohan) [inline swssconfig/sample/sample.json]: use another ip, e.g., 10.0.0.1

127.0.0.1 is local loopback address, should never appear on the wire whereas the decap is to decap the packet on the wire.

File: orchagent/tunneldecaporch.cpp [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+ */
+bool TunnelDecapOrch::addDecapTunnel(string type, IpAddresses dst_ip, string dscp, string ecn, string ttl)
+{
+
+    SWSS_LOG_ENTER();
+
+    int length_of_tunnel_attrs = 5;
+
+    // adding tunnel attributes to array and writing to ASIC_DB
+    sai_attribute_t tunnel_attrs[length_of_tunnel_attrs];
```
Comment (lguohan) [inline orchagent/tunneldecaporch.cpp]: use std::vector<sai_attribute_t> to accommodate variable number of attributes.

File: orchagent/tunneldecaporch.cpp [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    // decap ecn mode (copy from outer/standard)
+    tunnel_attrs[2].id = SAI_TUNNEL_ATTR_DECAP_ECN_MODE;
+    if (ecn == "copy_from_outer") 
+    {
+        tunnel_attrs[2].value.s32 = SAI_TUNNEL_DECAP_ECN_MODE_COPY_FROM_OUTER;
+    }
+    else if (ecn == "standard") 
+    {
+        tunnel_attrs[2].value.s32 = SAI_TUNNEL_DECAP_ECN_MODE_STANDARD;
+    }
```
Comment (lguohan) [inline orchagent/tunneldecaporch.cpp]: else {
   SWSS_LOG_ERROR("invalid tunnel decap ttl attribute %s")
}

File: orchagent/tunneldecaporch.cpp [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        string ttl_mode;
+        
+        if (op == SET_COMMAND)
+        {
+
+            for (auto i : kfvFieldsValues(t))
+            {
+                if (fvField(i) == "tunnel_type") 
+                {
+                    tunnel_type = fvValue(i);
```
Comment (lguohan) [inline orchagent/tunneldecaporch.cpp]: can we validate all parameters?

File: orchagent/tunneldecaporch.cpp [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        tunnel_attrs[4].value.s32 = SAI_TUNNEL_DSCP_UNIFORM_MODEL;
+    }
+    else if (dscp == "pipe") 
+    {
+        tunnel_attrs[4].value.s32 = SAI_TUNNEL_DSCP_PIPE_MODEL;
+    }
+
+    // write attributes to ASIC_DB
+    sai_object_id_t tunnel_id;
+    sai_status_t status = sai_tunnel_api->create_tunnel(&tunnel_id, length_of_tunnel_attrs, tunnel_attrs);
```
Comment (lguohan) [inline orchagent/tunneldecaporch.cpp]: what's if just change one attribute of existing tunnel? are we going to create a new one? can we maintain a tunnel cache and use set_attribute to change the attribute accordingly?

File: orchagent/tunneldecaporch.cpp [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+
+    sai_ip_address_t tunnel_dst_ip;
+    tunnel_dst_ip.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+
+    set<IpAddress> tunnel_ips = dst_ip.getIpAddresses();
+    struct sockaddr_in tunnel_ip_struct;
+    string ip;
+
+    // loop through the IP list and create a new tunnel table entry for every IP (in network byte order)
+    for (auto it = tunnel_ips.begin(); it != tunnel_ips.end(); ++it)
```
Comment (lguohan) [inline orchagent/tunneldecaporch.cpp]: consider the scenario to add or delete tunnel dst ip.

File: orchagent/tunneldecaporch.cpp [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    // TODO:
+    // check if the database is already populated with the correct data
+    // don't create duplicate entries
+    // there should also be "business logic" for netbouncer in the "tunnel application" code, which is a different source file and daemon process
+
+    // adding tunnel table entry attributes to array and writing to ASIC_DB
+    int length_of_tunnel_table_entry_attrs = 5;
+    sai_attribute_t tunnel_table_entry_attrs[length_of_tunnel_table_entry_attrs];
```
Comment (lguohan) [inline orchagent/tunneldecaporch.cpp]: use std::vector<sai_attribute_t>

File: orchagent/tunneldecaporch.cpp (hunk 11 above)
Comment (scsmncao) [inline orchagent/tunneldecaporch.cpp]: updated!

File: orchagent/tunneldecaporch.cpp [hunk 15]
```diff
@@ -0,0 +1,515 @@
...
+ *    @brief remove decap tunnel termination entry
+ *
+ * Arguments:
+ *    @param[in] key - key of the tunnel from APP_DB
+ *
+ * Return Values:
+ *    @return true on success and false if there's an error
+ */
+bool TunnelDecapOrch::removeDecapTunnelTermEntry(string key, TunnelTermEntry &tunnel_term_info)
+{
```
Comment (scsmncao) [inline orchagent/tunneldecaporch.cpp]: don't need tunnel_term_info, will fix!

File: orchagent/tunneldecaporch.cpp [hunk 16]
```diff
@@ -0,0 +1,515 @@
...
+    sai_status_t status;
```
Comment (scsmncao) [inline orchagent/tunneldecaporch.cpp]: don't need key, will fix!

**Pull request merged**